engines/sherlock/talk.cpp

```diff
@@ -0,0 +1,1772 @@
+		if (objNum != -1) {
+			Object &obj = scene._bgShapes[objNum];
+
+			if (obj._seqSize < MAX_TALK_SEQUENCES) {
+				warning("Tried to copy too few still frames");
+			} else {
+				for (uint idx = 0; idx < MAX_TALK_SEQUENCES; ++idx) {
+					obj._sequences[idx] = STILL_SEQUENCES[speaker][idx];
+					if (idx > 0 && !TALK_SEQUENCES[speaker][idx] && !TALK_SEQUENCES[speaker][idx - 1])
+						break;
+				}
+
+				obj._frameNumber = 0;
+				obj._seqTo = 0;
+			}
+		}
+	}
+}
+
+/**
+ * Parses a reply for control codes and display text. The found text is printed within
+ * the text window, handles delays, animations, and animating portraits.
+ */
+void Talk::doScript(const Common::String &script) {
+	Animation &anim = *_vm->_animation;
+	Events &events = *_vm->_events;
+	Inventory &inv = *_vm->_inventory;
+	Map &map = *_vm->_map;
+	People &people = *_vm->_people;
+	Scene &scene = *_vm->_scene;
+	Screen &screen = *_vm->_screen;
+	Sound &sound = *_vm->_sound;
+	UserInterface &ui = *_vm->_ui;
+	int wait = 0;
+	bool pauseFlag = false;
+	bool endStr = false;
+	int yp = CONTROLS_Y + 12;
+	int charCount = 0;
+	int line = 0;
+	bool noTextYet = true;
+	bool openTalkWindow = false;
+	int seqCount;
+
+	_savedSequences.clear();
+
+	const byte *scriptStart = (const byte *)script.c_str();
+	const byte *scriptEnd = scriptStart + script.size();
+	const byte *str = scriptStart;
+
+	if (_scriptMoreFlag) {
+		_scriptMoreFlag = 0;
+		str = scriptStart + _scriptSaveIndex;
+	}
+
+	// Check if the script begins with a Stealh Mode Active command
+	if (str[0] == STEALTH_MODE_ACTIVE || _talkStealth) {
+		_talkStealth = 2;
+		_speaker |= 128;
+	} else {
+		pushSequence(_speaker);
+		ui.clearWindow();
+
+		// Need to switch speakers?
+		if (str[0] == SWITCH_SPEAKER) {
+			_speaker = str[1] - 1;
+			str += 2;
+			pullSequence();
+			pushSequence(_speaker);
+			setSequence(_speaker);
+		} else {
+			setSequence(_speaker);
+		}
+
+		// Assign portrait location?
+		if (str[0] == ASSIGN_PORTRAIT_LOCATION) {
+			switch (str[1] & 15) {
+			case 1:
+				people._portraitSide = 20;
+				break;
+			case 2:
+				people._portraitSide = 220;
+				break;
+			case 3:
+				people._portraitSide = 120;
+				break;
+			default:
+				break;
+
+			}
+
+			if (str[1] > 15)
+				people._speakerFlip = true;
+			str += 2;
+		}
+
+		// Remove portrait?
+		if (str[0] == REMOVE_PORTRAIT) {
+			_speaker = 255;
+		} else {
+			// Nope, so set the first speaker
+			people.setTalking(_speaker);
+		}
+	}
+
+	do {
+		Common::String tempString;
+		wait = 0;
+
+		byte c = str[0];
+		if (!c) {
+			endStr = true;
+		} else if (c == '{') {
+			// Start of comment, so skip over it
+			while (*str++ != '}')
+				;
+		} else if (c >= 128) {
+			// Handle control code
+			switch (c) {
+			case SWITCH_SPEAKER:
+				// Save the current point in the script, since it might be intterupted by
+				// doing bg anims in the next call, so we need to know where to return to
+				_scriptCurrentIndex = str - scriptStart;
+
```

Comment:
I don't understand this variable \_scriptCurrentIndex.

It is set in many places, with a big comment above it, but it is only _used_ in a single place, and it is set right before it is used there.